src/transformStateWithClones.js

```diff
@@ -6,8 +6,40 @@
  *
  * @return {Object[]}
  */
+const ADD_PROPERTIES = 'addProperties';
+const REMOVE_PROPERTIES = 'removeProperties';
+const CLEAR = 'clear';
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  let tempState = { ...state };
+  const stateHistory = [];
+
+  for (const action of actions) {
+    const { type, extraData, keysToRemove } = action;
+
+    switch (type) {
+      case ADD_PROPERTIES:
+        tempState = {
+          ...tempState,
+          ...extraData,
+        };
+        break;
+
+      case REMOVE_PROPERTIES:
+        for (const key of keysToRemove) {
+          delete tempState[key];
+        }
+        break;
+
+      case CLEAR:
+        tempState = {};
+        break;
+    }
+
+    stateHistory.push({ ...tempState });
```

Comment:
[DONT REPEAT YOURSELF] - Instead of creating a new object copy inside each case, you could create a single copy after the switch statement, as the action of pushing to 'stateHistory' is repeated in every case.